src/annotator/host.js

```diff
@@ -0,0 +1,102 @@
+import $ from 'jquery';
+
+// TODO - Convert this to an ES import once the `Guest` class is converted to JS.
+// @ts-expect-error
+const Guest = require('./guest');
+
+export default class Host extends Guest {
+  constructor(element, config) {
+    // Some config settings are not JSON-stringifiable (e.g. JavaScript
+    // functions) and will be omitted when the config is JSON-stringified.
+    // Add a JSON-stringifiable option for each of these so that the sidebar can
+    // at least know whether the callback functions were provided or not.
+    if (config.services?.length > 0) {
+      const service = config.services[0];
+      if (service.onLoginRequest) {
+        service.onLoginRequestProvided = true;
+      }
+      if (service.onLogoutRequest) {
+        service.onLogoutRequestProvided = true;
+      }
+      if (service.onSignupRequest) {
+        service.onSignupRequestProvided = true;
+      }
+      if (service.onProfileRequest) {
+        service.onProfileRequestProvided = true;
+      }
+      if (service.onHelpRequest) {
+        service.onHelpRequestProvided = true;
+      }
+    }
+
+    // Make a copy of all config settings except `config.sidebarAppUrl`, the app base URL,
+    // and `config.pluginClasses`
+    const configParam =
+      'config=' +
+      encodeURIComponent(
+        JSON.stringify(
+          Object.assign({}, config, {
+            sidebarAppUrl: undefined,
+            pluginClasses: undefined,
+          })
+        )
+      );
+
+    const sidebarAppSrc = config.sidebarAppUrl + '#' + configParam;
+
+    // Create the iframe
+    const app = $('<iframe></iframe>')
+      .attr('name', 'hyp_sidebar_frame')
+      // enable media in annotations to be shown fullscreen
+      .attr('allowfullscreen', '')
+      .attr('seamless', '')
+      .attr('src', sidebarAppSrc)
+      .attr('title', 'Hypothesis annotation viewer')
+      .addClass('h-sidebar-iframe');
+
+    let externalContainer = null;
+
+    if (config.externalContainerSelector) {
+      // Use the native method to also validate the input
+      externalContainer = document.querySelector(
+        config.externalContainerSelector
+      );
+    }
+    super(...arguments);
```

Comment:
any reason this can't be the first statement in the ctor?

Reply:
I can't see any problem with doing so. I'll make this change.

Reply:
Update: There _is_ a reason that `super(...)` needs to be called part-way down the function. It turns out that the `super(...)` call results in the `BucketBar` constructor being called and that class in turn assumes that it is called after the bucket bar's container element (`.annotator-frame`) is in the DOM. This needs refactoring, but I'll do that as a separate task.